dsdl_compiler/README.md

```diff
@@ -0,0 +1,108 @@
+# UAVCAN DSDL compiler for the libcanard
+
+## Overview
+Libcanard_dsdlc is a tool to generate UAVCAN DSDL definitions to libcanard compatible message c-modules.
+
+Modules have: defines, enums, unions, structs, and encoding and decoding functions as defined in UAVCAN DSDL. Encoding and decoding functions use the decode and encode functions of libcanard for bit packing / unpacking.
+
+In c there is no namespace, so all the generated #define and function names are long having full folder path included. This is made to prevent the collision with each other and to the rest of the system.
+
+## Install & Integration
+To get libcanard from the git, make sure all the submodules are fetched too.
+
+`git submodules update --init --recursive`
+
+### Generating files
+`python3 libcanard_dsdlc --outdir <outdir> <dsdl-definition-uavcan-folder>`
+
+NOTE: If python2 is used, monotonic library is needed.
+
+`pip install monolith`
+
+### Using generated c-modules
+Include all or only selected message c-files to your compiler script (e.g. Makefile). Add include path to root of the <dsdl-generate-output-folder>.
+
+NOTE: compiled *.o files can't be compiled into to flat "build" directory as some files in DSDL have the same name.
```

Comment:
This is a major limitation, isn't it? May I suggest using full type names for generated files, e.g. `uavcan_protocol_NodeStatus.c`?

Reply:
Well not major.

1) as you can make Makefile to keep the folder structure also in the out directory. (This is what we mostly do)

If putting full type name to file names it becomes a bit ugly as type name is also in the path.
```cpp
#include <uavcan/protocol/uavcan_protocal_SoftwareVersion.h>
```
2) Libcanard is for resource limited uCs, most likely you will not need all the APIs at the same uC. (If you not using solution 1.)

How about putting some NOTE and e.g. example of Makefile into readme how to retain folder structure?

What do you recommend?

Reply:
The problem is that certain build systems (e.g. ChibiOS Makefiles) do not permit one to retain the original directory structure in the build output directory. So I see these solutions so far:

1. Use underscore-separated full data type path as file name for C files, while using short type names for header files: `uavcan/protocol/SoftwareVersion.h`, `uavcan/protocol/uavcan_protocal_SoftwareVersion.c`. This is exceptionally ugly though.

2. Put all entities in the header files, declare functions static. I understand that there may be some code size disadvantages involved, but are they significant enough, and is the link-time optimization not a viable option?

Reply:
I wonder if generated Makefile would work. So that it allows to build lib uavcan_dsdl.a or something like that?
If lib does not work out then 1. is the way to go.

Reply:
I don't think generating a Makefile is a good idea. People have all sorts of weird build process and flags and forcing one upon them will be annoying.

I think  number 1 is the best solution, although not very convenient I agree.

Reply:
> I don't think generating a Makefile is a good idea. People have all sorts of weird build process and flags and forcing one upon them will be annoying.

As a long time practitioner of weird build processes and flags, I couldn't agree more.

Can we please talk about the option 2? It seems like a no-brainer to me, but I could be missing something.

Reply:
It should be indicated here that this is just one of the available options, and that the other option is to generate header files, where the limitations do not apply. Also `*.o` should be enclosed in backticks, otherwise it breaks the Markdown formatting.